scarlet/psf.py

```diff
@@ -10,33 +13,79 @@ def moffat(coords, y0, x0, amplitude, alpha, beta=1.5):
     """
-    Y, X = coords
+    X, Y = np.meshgrid(x, y)
     return (amplitude*(1+((X-x0)**2+(Y-y0)**2)/alpha**2)**-beta)
 
 
-def gaussian(coords, y0, x0, amplitude, sigma):
+def gaussian(y, x, y0=0, x0=0, amplitude=None, sigma=1):
     """Circular Gaussian Function
+
+    Parameters
+    ----------
+    y: array
+        Coordinates to sample the circular gaussian in the y dimension.
+    x: array
+        Coordinates to sample the circular gaussian in the x dimension.
+    amplitude: float
+        Height of the gaussian. If `amplitude` is `None` then
+        `amplitude = 1/(np.pi**2*sigma**2)`
+        (so that the gaussian is normalized).
+    sigma: float
+        Standard deviation of the gaussian.
+
+    Returns
+    -------
+    result: array
+        A 2D circular gaussian sampled at the coordinates `(y_i, x_j)`
+        for all i in `y` and j in `x`.
     """
-    Y, X = coords
+    if amplitude is None:
+        amplitude = 1/(np.pi**2*sigma**2)
+    X, Y = np.meshgrid(x, y)
     return (amplitude*np.exp(-((X-x0)**2+(Y-y0)**2)/(2*sigma**2)))
 
 
-def double_gaussian(coords, y0, x0, A1, sigma1, A2, sigma2):
+def double_gaussian(y, x, y0=0, x0=0, A1=None, sigma1=1, A2=None, sigma2=1):
     """Sum of two Gaussian Functions
     """
-    return gaussian(coords, y0, x0, A1, sigma1) + gaussian(coords, y0, x0, A2, sigma2)
+    return gaussian(y, x, y0, x0, A1, sigma1) + gaussian(y, x, y0, x0, A2, sigma2)
 
 
-def generate_psf_image(func, shape, center=None, *args, **kwargs):
+def generate_psf_image(func, shape, subsamples=10, normalize=True, **kwargs):
     """Generate a PSF image based on a function and shape
+
+    This function uses a subdivided pixel scale to sample `func` at
+    higher frequencies and uses the trapezoid rule to estimate the integral
+    in each subsampled region. The subsampled pixel are then summed to
+    give the estimated integration values for each pixel at the scale requested
+    by `shape`.
+
+    Parameters
+    ----------
+    shape: tuple
+        Expected shape of the output image.
+    subsamples: int
+        Number of pixels to subdivide each output pixel for
+        a more accurate integration.
+    normalize: bool
+        Whether or not to normalize the PSF.
+    kwargs: keyword arguments
+        Keyword arguments to pass to `func`.
+
+    Returns
+    -------
+    result: array
+        Integrated image generated
     """
-    X = np.arange(shape[1])
-    Y = np.arange(shape[0])
-    X, Y = np.meshgrid(X, Y)
-    coords = np.stack([Y, X])
-    if center is None:
-        center = (shape[0]-1) // 2, (shape[1]-1) // 2
-    return func(coords, center[0], center[1], *args, **kwargs)
+    ry, rx = np.array(shape) // 2
+
+    y = np.linspace(-ry, ry, shape[0])
+    x = np.linspace(-rx, rx, shape[1])
+    f = partial(func, **kwargs)
+    result = apply_2D_trapezoid_rule(y, x, partial(f, **kwargs), subsamples)
+    if normalize:
+        result /= result.sum()
+    return result
```

Comment:
Should we not make all psfs Fourier objects?

Reply:
Hmmm, you make a good point. I can't seem to think of a time that we'll want to generate a PSF that isn't a `Fourier` object and even if such a case comes up we can just call `Fourier.image` to get the image array anyway. This will take a little time updating a lot of the docs and tests but I think it is the correct thing to do.

Reply:
There is not a BIG advantage to have psfs turned into `Fourier` object so if the change is a hastle, maybe  it's not worth it.
Though, it would make declaring analytic psfs slightly nicer on the users side, instead of having to write something like `generate_psf_image(analytic_psf)`, we would just declare the analytic psf as a Fourier object. I think it also makes sense conceptually, but that sounds like the kind of stylistic detail @pmelchior likes to weigh in.

Reply:
Ok, there is an implementation decision that needs to be made if we return a `Fourier` object. The issue is that `generate_psf_image` returns a 2D PSF image, but the model image is a PSF with an added dimension for the band. This is accomplished by adding an axis to the PSF image using either `psf[None]` or `psf[np.newaxis]` (they are the same thing, since apparently `np.newaxis` is `None`).

The problem is if we add a new axis to a `Fourier` object via slicing we need to decide the default behavior, is it an axis that will be transformed or will it be an axis the the FFT skips? Currently it is up to the user to decide since they have to first add the new axis to the PSF image and then create a `Fourier` object where they specify the `axes` used.

My first thought is that if we add a new axis it should *not* be transformed, assuming that most people wanting to add an axis are adding a band dimension. But I could be missing something, so I'm curious to hear your thoughts.

Reply:
I went with my solution to skip any new axes and fixed the tests accordingly. Have a look @herjy and if you and @pmelchior  approve this ticket is ready to be merged.